packages/graphqlgen/src/generators/ts-generator.ts

```diff
@@ -275,38 +469,90 @@ function renderResolverFunctionInterface(
 }
 
 function renderResolverTypeInterface(
-  type: GraphQLTypeObject,
+  type: GraphQLTypeObject | GraphQLInterfaceObject,
   modelMap: ModelMap,
+  interfacesMap: InterfacesMap,
+  unionsMap: UnionsMap,
   context?: ContextDefinition,
+  interfaceName: string = 'Type',
 ): string {
+  // TODO: Refactor this with proper union type branching once the
+  // type modelling is refactored itself to support that.
+  let extend = ''
+
+  if (!type.type.isInterface) {
+    type = type as GraphQLTypeObject
+
+    extend =
+      type.implements && type.implements.length
+        ? `extends ${type.implements
+            .map(typeInterface => `${typeInterface}Resolvers.InterfaceType`)
+            .join(',')}`
+        : ''
+  }
+
   return `
-  export interface Type {
+  export interface ${interfaceName} ${extend} {
```

Comment:
Due to this extends we run into type errors currently, example from the test suite:

```ts
export namespace ImageResolvers {
  export interface Type extends MediaResolvers.InterfaceType {
    id: (
      parent: Image,
      args: {},
      ctx: Context,
      info: GraphQLResolveInfo
    ) => string | Promise<string>;
//...
```
```ts
export namespace MediaResolvers {
  export interface InterfaceType {
    id: (
      parent: Image | Video,
      args: {},
      ctx: Context,
      info: GraphQLResolveInfo
    ) => string | Promise<string>;
//...
```

The issue is in the `parent` argument of resolvers.

The interface enforces for example in `id` field resolver to be `parent: Image | Video`.

But then the `ImageResolvers.Type['id']` has `parent: Image`.

This is a type error because ImageResolvers.Type['id'] isn't meeting the interface contract.

Possible solutions:

1. do not generate `export interface <Object> extends <Object>Resolvers.InterfaceType`
2. change the interface to emit a union at the property level:
```ts
    id:
      | ((
          parent: Image,
          args: {},
          ctx: Context,
          info: GraphQLResolveInfo,
        ) => string | Promise<string>)
      | ((
          parent: Video,
          args: {},
          ctx: Context,
          info: GraphQLResolveInfo,
        ) => string | Promise<string>)
```

Option 1 seems pretty good to right now because:

- its trivial to implement
- the generated `... extends ...` is questionable in the first place. The code is generated so what use is the contract in upholding type safety exactly?

Maybe option 2 would be something to revisit in future polishing work, but I'm not even sure about that.

Reply:
Another advantage to note of option 1 is that it allows simplifying the generator implementation and generation output because the use-case for the `InterfaceType` type is gone, example snapshot diff:

```diff
❯ gd src/tests/typescript/__snapshots__/basic.test.ts.snap
diff --git a/packages/graphqlgen/src/tests/typescript/__snapshots__/basic.test.ts.snap b/packages/graphqlgen/src/tests/typescript/__snapshots__/basic.test.ts.snap
index 8eab170..f47f6b2 100644
--- a/packages/graphqlgen/src/tests/typescript/__snapshots__/basic.test.ts.snap
+++ b/packages/graphqlgen/src/tests/typescript/__snapshots__/basic.test.ts.snap
@@ -448,7 +448,7 @@ export namespace ImageResolvers {
     info: GraphQLResolveInfo
   ) => Dimensions | Promise<Dimensions>;

-  export interface Type extends MediaResolvers.InterfaceType {
+  export interface Type {
     id: (
       parent: Image,
       args: {},
@@ -498,7 +498,7 @@ export namespace VideoResolvers {
     info: GraphQLResolveInfo
   ) => number | Promise<number>;

-  export interface Type extends MediaResolvers.InterfaceType {
+  export interface Type {
     id: (
       parent: Video,
       args: {},
@@ -525,22 +525,6 @@ export namespace VideoResolvers {
 }

 export namespace MediaResolvers {
-  export interface InterfaceType {
-    id: (
-      parent: Image | Video,
-      args: {},
-      ctx: Context,
-      info: GraphQLResolveInfo
-    ) => string | Promise<string>;
-
-    url: (
-      parent: Image | Video,
-      args: {},
-      ctx: Context,
-      info: GraphQLResolveInfo
-    ) => string | Promise<string>;
-  }
-
   export interface Type {
     __resolveType: GraphQLTypeResolver<Image | Video, Context>;
   }
```